File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/voice/AudioFocusDelegateProvider.java
package com.mapbox.navigation.ui.voice;

import android.media.AudioManager;
import android.os.Build;

import androidx.annotation.NonNull;

class AudioFocusDelegateProvider {

  @NonNull
  private final AudioFocusDelegate audioFocusDelegate;

  AudioFocusDelegateProvider(AudioManager audioManager) {
    audioFocusDelegate = buildAudioFocusDelegate(audioManager);
  }

  @NonNull
  AudioFocusDelegate retrieveAudioFocusDelegate() {
    return audioFocusDelegate;
  }

  @NonNull
  private AudioFocusDelegate buildAudioFocusDelegate(AudioManager audioManager) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      return new Api26AudioFocusDelegate(audioManager);
    }
    return new SpeechAudioFocusDelegate(audioManager);
  }
}

File: libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/AudioFocusDelegateProvider.kt
package com.mapbox.navigation.ui.voice.api

import android.content.Context
import android.media.AudioManager
import android.os.Build
import com.mapbox.navigation.ui.voice.options.VoiceInstructionsPlayerOptions

internal object AudioFocusDelegateProvider {

    fun defaultAudioFocusDelegate(
        context: Context,
        options: VoiceInstructionsPlayerOptions,
    ) = createAudioFocusDelegate(
        context.getSystemService(Context.AUDIO_SERVICE) as AudioManager,
        VoiceInstructionsPlayerAttributesProvider.retrievePlayerAttributes(options)
    )

    fun createAudioFocusDelegate(
        audioManager: AudioManager,
        playerAttributes: VoiceInstructionsPlayerAttributes,
    ): AsyncAudioFocusDelegate {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            OreoAndLaterAudioFocusDelegate(audioManager, playerAttributes)
        } else {
            PreOreoAudioFocusDelegate(audioManager, playerAttributes)
        }
    }
}
